improve error handling for matching annotations

byxorna opened the pull request:
# What and why?

This PR improves flow of matching requested sidecar configurations in the mutation handler. A bit of refactoring was necessary to make error conditions more clear, and map those failures to clear reasons in exposed metrics. I have also beefed up tests some to cover all the error cases the mutation route can get into.

# Testing Steps

`make test all`

# Reviewers

Required reviewers: @defect @alex-laties @komapa @roymarantz @gtorre 
Request reviews from other people you want to review this PR in the "Reviewers" section on the right.

> :warning: this PR must have at least 2 thumbs from the [MAINTAINERS.md](/MAINTAINERS.md) of the project before merging!

kevinsteck commented on pkg/server/webhook.go:
```diff
@@ -134,20 +134,29 @@ func (whsvr *WebhookServer) requiredMutation(ignoredList []string, metadata *met
+	requestAnnotationKey := whsvr.Config.AnnotationNamespace + "/request"
+
+	status, ok := annotations[statusAnnotationKey]
+	if ok && strings.ToLower(status) == "injected" {
```
Should `"injected"` be stored in a const?

kevinsteck commented on pkg/server/webhook.go:
```diff
@@ -134,20 +134,29 @@ func (whsvr *WebhookServer) requiredMutation(ignoredList []string, metadata *met
-			return ""
-		}
+	requestedInjection, ok := annotations[requestAnnotationKey]
+	injectionKey := strings.ToLower(requestedInjection)
```
Might move this below the `!ok` for clarity.

kevinsteck commented on pkg/server/webhook.go:
```diff
@@ -134,20 +134,29 @@ func (whsvr *WebhookServer) requiredMutation(ignoredList []string, metadata *met
+	injectionKey := strings.ToLower(requestedInjection)
+	if !ok {
+		glog.Infof("Pod %s/%s annotation %s is missing, skipping injection.", metadata.Namespace, metadata.Name, statusAnnotationKey)
+		return requestedInjection, ErrMissingRequestAnnotation
```
Might want to `return "", ErrMissingRequestAnnotation` for clarity as well.

byxorna replied:
Oops, good catch!

kevinsteck commented on pkg/server/webhook.go:
```diff
@@ -134,20 +134,29 @@ func (whsvr *WebhookServer) requiredMutation(ignoredList []string, metadata *met
+	}
+	if !whsvr.Config.HasInjectionConfig(requestedInjection) {
+		glog.Errorf("Mutation policy for %v/%v: status:%q requested injection: %s was not in configuration, skipping", metadata.Namespace, metadata.Name, status, requestedInjection)
+		return requestedInjection, ErrRequestedSidecarNotFound
```
Same as above.

byxorna replied:
because we parse out the requested injection in this function, the caller doesnt actually know what the requested injection is. To me, it makes sense to return both the name of the injection config the pod requested, as well as the error indicating that request cannot be satisfied

byxorna commented:
Thanks for the feedback @kevinsteck! ive addressed comments in the most recent commit https://github.com/tumblr/k8s-sidecar-injector/pull/4/commits/9a4c4f626012269f7a244b97f3138baf39cf5ba2.

alex-laties commented on internal/pkg/config/config.go:
```diff
@@ -64,15 +65,16 @@ func (c *Config) ReplaceInjectionConfigs(replacementConfigs []InjectionConfig) {
 	defer c.RUnlock()
-	i, ok := c.Injections[key]
+	k := strings.ToLower(key)
+	i, ok := c.Injections[k]
```
note that `i` here would be a copy of `c.Injections[k]`, not a reference, so if `c.Injections[k].(Name|Containers|Volumes|Environment|VolumeMounts)` ever change value, `i` will not get updated, and the returned pointer from this function would essentially be stale. Either the docs need to reflect this or the functionality needs to be adjusted.

byxorna replied:
Very good catch! I am refactoring this to use pointers, so this should be ameliorated shortly. See latest commit

byxorna commented on pkg/server/webhook.go:
```diff
@@ -24,6 +24,9 @@ var (
 	deserializer  = codecs.UniversalDeserializer()
 
+	// StatusInjected is the annotation value for /status that indicates an injection was already performed on this pod
+	StatusInjected = "injected"
```
excellent catch. ill update to a const

byxorna commented on pkg/server/webhook.go:
```diff
@@ -364,22 +376,35 @@ func (whsvr *WebhookServer) mutate(ar *v1beta1.AdmissionReview) *v1beta1.Admissi
+	if err != nil {
+		glog.Infof("Skipping mutation of %s/%s: %v", pod.Namespace, pod.Name, err)
+		var reason string
+		switch err {
```
good call. changeset incoming

byxorna commented:
@alex-laties thanks for the review! I updated the silent chain in the config loader to use references now, so I think the issue you mentioned should be fixed.
